File: clients/swift/Sources/messaging-cli/OfflineQueue.swift
import Foundation
import SQLite

class OfflineQueue {
    private var db: Connection!

    private let table = Table("queued_messages")
    private let colLocalId = Expression<String>("local_id")
    private let colServerId = Expression<String?>("server_id")
    private let colToUser = Expression<String>("to_user")
    private let colText = Expression<String>("text")
    private let colQueuedAt = Expression<Int64>("queued_at")
    private let colStatus = Expression<String>("status")

    init() {
        let dir = FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent(".messaging-cli")
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let dbPath = dir.appendingPathComponent("queue.db").path
        db = try! Connection(dbPath)
        createTable()
    }

    private func createTable() {
        try! db.run(table.create(ifNotExists: true) { t in
            t.column(colLocalId, primaryKey: true)
            t.column(colServerId)
            t.column(colToUser)
            t.column(colText)
            t.column(colQueuedAt)
            t.column(colStatus, defaultValue: "pending")
        })
    }

    func enqueue(to: String, text: String) -> String {
        let localId = UUID().uuidString
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        try! db.run(table.insert(
            colLocalId <- localId,
            colServerId <- nil,
            colToUser <- to,
            colText <- text,
            colQueuedAt <- now,
            colStatus <- QueuedStatus.pending.rawValue
        ))
        return localId
    }

    func nextPending() -> QueuedMessage? {
        let query = table
            .filter(colStatus == QueuedStatus.pending.rawValue)
            .order(colQueuedAt.asc)
            .limit(1)
        guard let row = try? db.pluck(query) else { return nil }
        return QueuedMessage(
            localId: row[colLocalId],
            serverId: row[colServerId],
            toUser: row[colToUser],
            text: row[colText],
            queuedAt: row[colQueuedAt],
            status: QueuedStatus(rawValue: row[colStatus]) ?? .pending
        )
    }

    func markSending(localId: String) {
        let row = table.filter(colLocalId == localId)
        try? db.run(row.update(colStatus <- QueuedStatus.sending.rawValue))
    }

    func markSent(localId: String, serverId: String) {
        let row = table.filter(colLocalId == localId)
        try? db.run(row.update(colStatus <- QueuedStatus.sent.rawValue, colServerId <- serverId))
    }

    func markFailed(localId: String) {
        let row = table.filter(colLocalId == localId)
        try? db.run(row.update(colStatus <- QueuedStatus.failed.rawValue))
    }

    func markPending(localId: String) {
        let row = table.filter(colLocalId == localId)
        try? db.run(row.update(colStatus <- QueuedStatus.pending.rawValue))
    }

    func isDuplicate(serverId: String) -> Bool {
        let query = table.filter(colServerId == serverId)
        return ((try? db.scalar(query.count)) ?? 0) > 0
    }
}

File: clients/kotlin/src/main/kotlin/com/messaging/OfflineQueue.kt
package com.messaging

import org.jetbrains.exposed.sql.*
import org.jetbrains.exposed.sql.transactions.transaction
import java.io.File
import java.util.UUID

object QueuedMessages : Table("queued_messages") {
    val localId = varchar("local_id", 36)
    val serverId = varchar("server_id", 36).nullable()
    val toUser = varchar("to_user", 255)
    val text = text("text")
    val queuedAt = long("queued_at")
    val status = varchar("status", 20).default("pending")

    override val primaryKey = PrimaryKey(localId)
}

class OfflineQueue {
    init {
        val dir = File(System.getProperty("user.home"), ".messaging-cli")
        dir.mkdirs()
        val dbPath = File(dir, "queue.db").absolutePath
        Database.connect("jdbc:sqlite:$dbPath", "org.sqlite.JDBC")
        transaction {
            SchemaUtils.create(QueuedMessages)
        }
    }

    fun enqueue(to: String, text: String): String {
        val localId = UUID.randomUUID().toString()
        transaction {
            QueuedMessages.insert {
                it[QueuedMessages.localId] = localId
                it[toUser] = to
                it[QueuedMessages.text] = text
                it[queuedAt] = System.currentTimeMillis()
                it[status] = "pending"
            }
        }
        return localId
    }

    fun nextPending(): QueuedMessage? = transaction {
        QueuedMessages
            .select { QueuedMessages.status eq "pending" }
            .orderBy(QueuedMessages.queuedAt to SortOrder.ASC)
            .limit(1)
            .firstOrNull()
            ?.toQueuedMessage()
    }

    fun markSending(localId: String) = transaction {
        QueuedMessages.update({ QueuedMessages.localId eq localId }) {
            it[status] = "sending"
        }
    }

    fun markSent(localId: String, serverId: String) = transaction {
        QueuedMessages.update({ QueuedMessages.localId eq localId }) {
            it[QueuedMessages.serverId] = serverId
            it[status] = "sent"
        }
    }

    fun markFailed(localId: String) = transaction {
        QueuedMessages.update({ QueuedMessages.localId eq localId }) {
            it[status] = "failed"
        }
    }

    fun markPending(localId: String) = transaction {
        QueuedMessages.update({ QueuedMessages.localId eq localId }) {
            it[status] = "pending"
        }
    }

    fun isDuplicate(serverId: String): Boolean = transaction {
        QueuedMessages
            .select { QueuedMessages.serverId eq serverId }
            .count() > 0
    }

    private fun ResultRow.toQueuedMessage() = QueuedMessage(
        localId = this[QueuedMessages.localId],
        serverId = this[QueuedMessages.serverId],
        toUser = this[QueuedMessages.toUser],
        text = this[QueuedMessages.text],
        queuedAt = this[QueuedMessages.queuedAt],
        status = MessageStatus.valueOf(this[QueuedMessages.status].uppercase())
    )
}
